aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -1399,90 +1415,90 @@ TORCH_IMPL_FUNC(_convert_indices_from_csr_to_coo_structured_cpu)
  * https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/sparse/sparsetools/csr.h
  * Modified to ensure sorted BSR column indices.
  */
-template <class I, class T>
-void _csr_to_block_csr_cpu_kernel(
-    const I n_row,
-    const I n_col,
-    const I R,
+template <class I, class T, bool compressed_rows>
+void _compressed_to_block_compressed_cpu_kernel(
+    const I n_dim_compressed,
+    const I n_dim_plain,
     const I C,
-    const I* input_crow_indices,
-    const I* input_col_indices,
+    const I P,
+    const I D,
+    const I* input_compressed_indices,
+    const I* input_plain_indices,
     const T* input_values,
-    I* result_crow_indices,
-    I* result_col_indices,
+    I* result_compressed_indices,
+    I* result_plain_indices,
     T* result_values) {
-  // All blocks are possible, that is, may be allocated if a single non-zero
-  // value lives within them. Otherwise they're not.
-
-  // Allocate pointers for all possible column blocks plus 1
-  std::vector<T*> blocks(n_col / C + 1, (T*)0);
+  // All blocks are possible, that is, may be allocated if a single
+  // non-zero value lives within them. Otherwise they're not.
 
-  assert(n_row % R == 0);
-  assert(n_col % C == 0);
+  // Allocate pointers for all possible plain blocks plus 1
+  std::vector<T*> blocks(n_dim_plain / P + 1, (T*)0);
 
-  // Major assumptions
-  // 1. Blocks must be square
+  assert(n_dim_compressed % C == 0);
+  assert(n_dim_plain % P == 0);
 
-  // Number of blocks along rows
-  I n_brow = n_row / R;
-  // Number of blocks along columns
-  I n_bcol = n_col / C;
+  // Number of blocks along compressed dim
+  I n_bcompressed = n_dim_compressed / C;
+  // Number of blocks along plain_dim
+  I n_bplain = n_dim_plain / P;
 
   // Number of elements per block
-  I RC = R * C;
+  I CPD = C * P * D;
   // Number of blocks overall
   I n_blks = 0;
 
-  result_crow_indices[0] = 0;
-
-  // Iterate over blocks along rows
-  for (I block_i = 0; block_i < n_brow; block_i++) {
-    // Iterate over blocks along columns to locate non-zero blocks,
-    // this guarantees sorted block-column indices
-    for (I block_j = 0; block_j < n_bcol; block_j ++) {
-      for (I jj = input_crow_indices[R * block_i]; jj < input_crow_indices[R * (block_i + 1)]; jj++) {
-        I j = input_col_indices[jj]; // column index
-        if (j / C == block_j) {
-          blocks[block_j] = result_values + RC * n_blks;
-          result_col_indices[n_blks] = block_j;
+  result_compressed_indices[0] = 0;
+
+  // Iterate over blocks along compressed dim
+  for (I block_c = 0; block_c < n_bcompressed; block_c++) {
+    // Iterate over blocks along plain dim to locate non-zero blocks,
+    // this guarantees sorted plain dim indices
+    for (I block_p = 0; block_p < n_bplain; block_p ++) {
+      for (I i = input_compressed_indices[C * block_c]; i < input_compressed_indices[C * (block_c + 1)]; i++) {
```

Comment:
I suppose at least one of these `for` loops could be run in parallel?

Reply:
This whole code is ugly for its copying to CPU memory and back, and doing work through loops.  But indeed we could at least add some parallelization here.  What is the preferred way to to it for CPU code in PyTorch, would just going for `#pragma omp` be acceptable?

Reply:
@alexsamardzic , take a look at `at::parallel_for`.

Reply:
Thanks for the pointer!  Looked into it, there is a data dependency between loops, about `n_blks` variable so it's not trivially parallelizable, so let's keep it as is for now.